static/app/utils/useTeamsById.tsx

```diff
@@ -0,0 +1,112 @@
+import {useEffect, useMemo} from 'react';
+
+import OrganizationStore from 'sentry/stores/organizationStore';
+import TeamStore from 'sentry/stores/teamStore';
+import {useLegacyStore} from 'sentry/stores/useLegacyStore';
+import type {Team} from 'sentry/types';
+import {ApiQueryKey, useApiQuery} from 'sentry/utils/queryClient';
+
+interface UseTeamsById {
+  ids: string[];
+}
+
+interface UseTeamsBySlug {
+  slugs: string[];
+}
+
+interface UseAllTeams {}
+
+/**
+ * ids[] and slugs[] are split to ensure that only one is provided
+ */
+type UseTeamOptions = UseTeamsById | UseTeamsBySlug | UseAllTeams;
+
+interface UseTeamsResult {
+  isError: boolean | null;
+  isLoading: boolean;
+  teams: Team[];
+}
+
+type TeamQuery = [field: string, ids: string[]];
+
+function buildTeamsQueryKey(orgSlug: string, teamQuery: TeamQuery | null): ApiQueryKey {
+  const query: {query?: string} = {};
+
+  if (teamQuery && teamQuery[1].length) {
+    query.query = `${teamQuery[0]}:${teamQuery[1].join(',')}`;
+  }
+
+  return [`/organizations/${orgSlug}/teams/`, {query}];
+}
+
+/**
+ * @returns a tuple of the identifier field and the list of identifiers
+ */
+function getQueryFromOptions(options: UseTeamOptions): TeamQuery | null {
+  if ('ids' in options) {
+    if (!options.ids.length) {
+      return null;
+    }
+    return ['id', options.ids];
+  }
+
+  if ('slugs' in options) {
+    if (!options.slugs.length) {
+      return null;
+    }
+    return ['slug', options.slugs];
+  }
+
+  return null;
+}
+
+/**
+ * example usage:
+ * ```ts
+ * const {teams, isLoading, isError} = useTeamsV2({slugs: ['project-slug']});
+ * ```
+ */
+export function useTeamsById(options: UseTeamOptions = {}): UseTeamsResult {
+  const {organization} = useLegacyStore(OrganizationStore);
+  const storeState = useLegacyStore(TeamStore);
+
+  const query = useMemo<TeamQuery | null>(() => getQueryFromOptions(options), [options]);
+  const missingIds = query
+    ? query[1].filter(id => !storeState.teams.find(team => team[query[0]] === id))
+    : [];
+
+  // Wait until the store has loaded to start fetching
+  const shouldConsiderFetching = !!organization?.slug && !storeState.loading;
+  // Only fetch if there are missing teams
+  const hasMissing = missingIds.length > 0;
+  const queryEnabled = shouldConsiderFetching && hasMissing;
+
+  const queryKey = buildTeamsQueryKey(organization?.slug ?? '', query);
+  const {
+    data: additionalTeams = [],
+    isLoading,
+    isError,
+  } = useApiQuery<Team[]>(queryKey, {
+    staleTime: 0,
+    enabled: queryEnabled,
+  });
```

Comment:
I wonder if we need to disable retries. Is there any reason this would give a 404 or some other kind of expected 4xx error?

Reply:
retry seems fine for now

Reply:
maybe 404 if the team doesn't exist